documentation/writing_tests.md

```diff
@@ -0,0 +1,949 @@
+
+Let's define an example in which a SDK client might need several lines of setup in two specific tests:
+
+```ts
+describe("some group of functionalities", function() {
+  let client;
+
+  beforeEach(function() {
+    client = new Client();
+  });
+
+  it("should test A with a proper setup", function() {
+    client.prepare({
+      propertyA: 1,
+      propertyB: 2,
+      propertyC: 3,
+      propertyD: 4,
+    });
+
+    const result = await client.A();
+    assert.ok(result.value);
+  });
+
+  it("should test B with a proper setup", function() {
+    client.prepare({
+      propertyA: 1,
+      propertyB: 2,
+      propertyC: 3,
+      propertyD: 4,
+    });
+
+    const result = await client.B();
+    assert.ok(result.value);
+  });
+});
+```
+
+It is possible to move the common setup code to a function in the same file, as in the following example code:
+
+```ts
+describe("some group of functionalities", function() {
+  let client: Client;
+
+  beforeEach(function() {
+    client = new Client();
+  });
+
+  function defaultPrepareClient(client: Client) {
+    client.prepare({
+      propertyA: 1,
+      propertyB: 2,
+      propertyC: 3,
+      propertyD: 4,
+    });
+  }
+
+  it("should test A with a proper setup", function() {
+    defaultPrepareClient(client);
+    const result = await client.A();
+    assert.ok(result.value);
+  });
+
+  it("should test B with a proper setup", function() {
+    defaultPrepareClient(client);
+    const result = await client.B();
+    assert.ok(result.value);
+  });
+});
+```
+
+### Universal utilities
+
+In case that a specific utility function might be useful for more than one test case, and more than one test file, and even potentially in more than one project, this code should be moved away from each client into a common private package in the [test-utils](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/test-utils/) folder. If the package doesn't exist, it should be created.
+
+Let's say we have a test utility that we call `retry.ts`, with the following content:
+
+```ts
+import { delay as coreDelay } from "@azure/core-http";
+
+/**
+ * A simple abstraction to retry, and exponentially de-escalate retrying, a
+ * given async function until it is fulfilled.
+ * @param {() => Promise<T>} target The async function you want to retry
+ * @param {number} delay The delay between each retry, defaults to 1000
+ * @param {number} timeout Maximum time we'll let this lapse before we quit retrying, defaults to Infinity
+ * @param {number} increaseFactor Increase factor of each retry, defaults to 1
+ * @returns {Promise<any>} Resolved promise
+ */
+export async function retry<T>(
+  target: () => Promise<T>,
+  delay: number = 1000,
```

Comment:
Maybe just name this `delayInMs` (like we force in our API anyways) and then you don't have to rename `delay` to `coreDelay`

Reply:
I will take your advice here!

Reply:
Done!